src/replication.c

```diff
@@ -4078,14 +4078,17 @@ void replicationSetPrimary(char *ip, int port, int full_sync_required) {
         server.primary->flag.dont_cache_primary = full_sync_required;
         freeClient(server.primary);
     }
-    disconnectAllBlockedClients(); /* Clients blocked in primary, now replica. */
 
     /* Setting primary_host only after the call to freeClient since it calls
      * replicationHandlePrimaryDisconnection which can trigger a re-connect
      * directly from within that call. */
     server.primary_host = sdsnew(ip);
     server.primary_port = port;
 
+    if (!server.cluster_enabled) {
```

Comment:
minor: can we add some comment above stating why this HAS to come after we set the new primary host (since unblocking clients might require redirecting to the new primary)

BTW - I think in the current implementation clients blocked on xread are also disco. we might consider changing this in a separate PR